src/rust/engine/fs/src/store.rs

```diff
@@ -306,6 +318,71 @@ impl Store {
       .to_boxed()
   }
 
+  ///
+  /// Downloads directories from remote CAS recursively to the local one.
+  ///
+  pub fn ensure_local_has_recursive(&self, digests: Vec<Digest>) -> BoxFuture<(), String> {
```

Comment:
This looks good :) And thanks in particular for the clarifying comments!

I suspect we may actually want the signature of this function to be: `ensure_local_has_recursive_directory(&self, digest: Digest) -> BoxFuture<(), String>` which is guaranteed to only be called with a `Digest` of a `Directory` (which we are guaranteed we'll be given by the remote execution API); that way, we don't need to determine the types of things; we know the top-level thing is a `Directory`, and then once we've retrieved that, its children are already strongly typed as either files or directories.

Sorry for misleading you in the first place!